### ISSUE TITLE
Unexpected emissions

### ISSUE THREAD

**Issue by PaulWoitaschek** [cashapp/turbine#112]
Using kotlin 1.6.21, coroutines 1.6.1 and turbine 0.8.0 I noticed the following behavior:

```kotlin
  private fun contentFlow(): Flow<Int> {
    return combine(
      flowOf(Unit),
      flowOf(2).onStart { emit(1) }
    ) { _, b ->
      b
    }.onEach { println(it) }
  }

  @Test
  fun turbineCollect() = runTest {
    contentFlow()
      .test {
        check(awaitItem() == 1)
        cancelAndIgnoreRemainingEvents()
      }
  }

  @Test
  fun regularCollect() = runTest {
    check(contentFlow().toList() == listOf(1, 2))
  }
```

This `turbineCollect` test case fails as it only emits the item `2`. The `regularCollect` behaves as expected and collects both the `1` and the `2`.

**Comment by JakeWharton** [comment]
Here's two interesting examples which highlight the problem:
```kotlin
    launch(Unconfined, start = UNDISPATCHED) {
      contentFlow().collect {
        println(it)
      }
    }
    launch(start = UNDISPATCHED) {
      contentFlow().collect {
        println(it)
      }
    }
```
Turbine uses the `Unconfined` dispatcher to ensure that it immediately can run to record events. This works great when dealing with normal dispatchers. The esoteric non-real behavior of`runTest` breaks that which is wild.

**Comment by JakeWharton** [comment]
This is a problem with the `combine` operator an `Unconfined` dispatcher that is entirely outside our control.

Here is an isolated test that you can file upstream on kotlinx.coroutines if you want:
```kotlin
@OptIn(ExperimentalCoroutinesApi::class)
class UnconfinedTest {
  @Test fun works() = runTest { testUnconfinedIn(this, EmptyCoroutineContext) }
  @Test fun broken() = runTest { testUnconfinedIn(this, Unconfined) }

  private suspend fun testUnconfinedIn(scope: CoroutineScope, collectContext: CoroutineContext) {
    val flow = combine(
      flowOf(Unit).onEach { println("A: $it") },
      flowOf(2).onStart { emit(1) }.onEach { println("B: $it") },
    ) { a, b ->
      println("A+B: $a $b")
      b
    }.onEach {
      println("C: $it")
    }

    val seen = Channel<Int>(UNLIMITED)
    val collectJob = scope.launch(collectContext, UNDISPATCHED) {
      flow.collect {
        seen.trySend(it).getOrThrow()
      }
    }

    assertEquals(1, seen.receive())
    assertEquals(2, seen.receive())

    collectJob.cancel()
  }
}
```

We have no power to fix this, and as #138 illustrates the use of the `Unconfined` dispatcher is more advantageous than not.
